### PR TITLE
Orm hibernate

### PR DESCRIPTION
**Pull request by maratb3k**

### CODE REVIEW

File: src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+        this.entityManager = entityManager;
+        this.userDao = userDao;
+        this.trainingDao = trainingDao;
+        this.trainerDao = trainerDao;
+    }
+
+    @Override
+    public boolean save(Trainee trainee) {
+        try {
+            System.out.println("rep");
```
Comment (SiarheiKatuzhenets) [inline src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java]: clean up the code - remove this kind of output everywhere

File: src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+
+    public TraineeDaoImpl(EntityManager entityManager, UserDaoImpl userDao, TrainingDaoImpl trainingDao, TrainerDaoImpl trainerDao) {
+        this.entityManager = entityManager;
+        this.userDao = userDao;
+        this.trainingDao = trainingDao;
+        this.trainerDao = trainerDao;
+    }
+
+    @Override
+    public boolean save(Trainee trainee) {
```
Comment (SiarheiKatuzhenets) [inline src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java]: for save method it is good to return saved entity - You have auto generated Id and how You will know which Id was generated?

File: src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+            System.out.println(e);
+            log.error("Error saving trainee: {}", trainee, e);
+            throw new DaoException("Error saving trainee: " + trainee, e);
+        }
+    }
+
+    @Override
+    public boolean addTrainer(Long traineeId, Long trainerId) {
+        try {
+            Trainee trainee = entityManager.find(Trainee.class, traineeId);
```
Comment (SiarheiKatuzhenets) [inline src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java]: move this logic from DAO to Service - It is not the responsibility of this class. In Service you should find trainee and trainer by Id and call DAO.update()

File: src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+        } catch (Exception e) {
+            log.error("Error adding trainer with id {} to trainee with id: {}", trainerId, traineeId, e);
+            throw new DaoException("Error adding trainer with id " + trainerId + " to trainee with id " + traineeId, e);
+        }
+    }
+
+    @Override
+    public boolean addTraining(Long traineeId, Long trainingId) {
+        try {
+            Trainee trainee = entityManager.find(Trainee.class, traineeId);
```
Comment (SiarheiKatuzhenets) [inline src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java]: also this logic should be in the service

File: src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+        }
+
+    }
+
+    @Override
+    public boolean update(Long id, Trainee updatedTrainee) {
+        try {
+            Trainee existingTrainee = entityManager.find(Trainee.class, id);
+            if(existingTrainee != null) {
+                if(updatedTrainee.getDateOfBirth() != null) {
```
Comment (SiarheiKatuzhenets) [inline src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java]: this validation should be in the service class. Here it must be only updating entity in database

File: src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+                return false;
+            }
+        } catch (Exception e) {
+            log.error("Error updating trainee with id: {}", id, e);
+            throw new DaoException("Error updating trainee with id " + id, e);
+        }
+    }
+
+    @Override
+    public boolean updatePassword(Long id, String newPassword) {
```
Comment (SiarheiKatuzhenets) [inline src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java]: I would like to have only one update method in the DAO. And Service class should contains: DAO.findById, then DAO.update entity

File: src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+            log.error("Unexpected error occurred while updating password for trainee with id: {}", id, e);
+            throw new DaoException("Unexpected error updating user password for trainee with id " + id, e);
+        }
+    }
+
+    @Override
+    public boolean updateTrainersList(Long id, Set<Trainer> trainers) {
+        try {
+            Trainee trainee = entityManager.find(Trainee.class, id);
+            if (trainee != null) {
```
Comment (SiarheiKatuzhenets) [inline src/main/java/com/example/crm_gym/repository/TraineeDaoImpl.java]: the same - move this kind of logic to Service class

File: src/main/java/com/example/crm_gym/repository/TrainingDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+            log.error("Error saving training: {}", training, e);
+            throw new DaoException("Error saving training " + training, e);
+        }
+    }
+
+    @Override
+    public boolean update(Long id, Training updatedTraining) {
+        try {
+            Training existingTraining = entityManager.find(Training.class, id);
+            if (existingTraining != null) {
```
Comment (SiarheiKatuzhenets) [inline src/main/java/com/example/crm_gym/repository/TrainingDaoImpl.java]: the same - move this logic to service from every DAO class

File: src/main/java/com/example/crm_gym/repository/UserDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    @Override
+    public boolean checkUsernameAndPassword(String username, String password) {
+        try {
+            String jpql = "SELECT u FROM User u WHERE u.username = :username AND u.password = :password";
+            User user = entityManager.createQuery(jpql, User.class)
+                    .setParameter("username", username)
+                    .setParameter("password", password)
+                    .getSingleResult();
```
Comment (SiarheiKatuzhenets) [inline src/main/java/com/example/crm_gym/repository/UserDaoImpl.java:130]: getSingleResult throw exception if entity not found. please make correct handling of this case

File: src/main/java/com/example/crm_gym/services/TraineeService.java [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+
+    @Autowired
+    public TraineeService(TraineeDAO traineeDAO, UserDAO userDAO) {
+        this.traineeDAO = traineeDAO;
+        this.userDAO = userDAO;
+    }
+
+    public boolean create(Date dateOfBirth, String address) {
+        try {
+            System.out.println("ddd");
```
Comment (SiarheiKatuzhenets) [inline src/main/java/com/example/crm_gym/services/TraineeService.java]: clean up the code

File: src/main/java/com/example/crm_gym/services/TrainingService.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+            return trainingDAO.save(training);
+        } catch (DaoException e) {
+            log.error("Error creating training", e);
+            return false;
+        }
+    }
+
+    public boolean update(String username, String password, Long id, Training training) {
+        try {
+            if (!userDAO.checkUsernameAndPassword(username, password)) {
```
Comment (SiarheiKatuzhenets) [inline src/main/java/com/example/crm_gym/services/TrainingService.java]: you have repeated code. Read about DRY principle. You can create separate method for that

File: src/main/java/com/example/crm_gym/services/TrainingTypeService.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+            if (!userDAO.checkUsernameAndPassword(username, password)) {
+                log.error("Invalid credentials for user {}", username);
+                return Collections.emptyList();
+            }
+            Optional<List<TrainingType>> trainingTypes = trainingTypeDAO.findAll();
+            if (trainingTypes.isPresent()) {
+                return trainingTypes.get();
+            } else {
+                log.warn("No training types found.");
+                return Collections.emptyList();
```
Comment (SiarheiKatuzhenets) [inline src/main/java/com/example/crm_gym/services/TrainingTypeService.java]: it is not good manner to have several 'return' in one method.

**Comment by maratb3k** [comment]
-

**Pull request closed without merging**